### PR TITLE
Shortened Timespans (again)

### PR DESCRIPTION
**Pull request by Asleeepp**
### Description

This PR aims to add an option to provide a shortened version of a Timespan.
Unlike my previous PR of this exact thing, this uses the lang file, and does not require the Command context, as I felt making this exclusive to commands was a bit silly.
Not sure if this will require a test, or how I could even make a test for this. If it does, please tell me, (and some guidance would be helpful). and I will make one.

---
**Target Minecraft Versions:** any 
**Requirements:** none 
**Related Issues:** none

### CODE REVIEW

**Comment by Asleeepp** [comment]
Oh, lovely, the tests are failing, guess i screwed something up

File: src/main/java/ch/njol/skript/util/Timespan.java [hunk 1]
```diff
@@ -159,31 +176,67 @@ else if (length == 3 && !hasMs || length == 4) // HH:MM:SS[.ms]
...
+	 * @return A {@link Timespan} object representing the parsed timespan, or {@code null} if the input string is invalid.
+	 */
+	@Nullable
+	public static Timespan parseShortenedTimespan(String value) {
+		if (value.isEmpty()) {
+			return null;
+		}
+
+		long totalMillis = 0;
+		Pattern pattern = Pattern.compile("(\\d+)([a-zA-Z]+)");
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/util/Timespan.java:216]: why are we compiling this every time the method is called?

File: src/main/java/ch/njol/skript/util/Timespan.java [hunk 2]
```diff
@@ -159,31 +176,67 @@ else if (length == 3 && !hasMs || length == 4) // HH:MM:SS[.ms]
...
 	}
-	
+
 	public Timespan() {
 		millis = 0;
 	}
 
+	/**
+	 * Parses a string representing a shortened timespan, such as "1d" for 1 day,
+	 * "2h" for 2 hours, or "30m" for 30 minutes.
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/util/Timespan.java:205]: These values aren't consistent. Mention that they're based on lang files.

File: src/main/java/ch/njol/skript/util/Timespan.java (hunk 1 above)
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/util/Timespan.java:216]: this doesn't support `1.4s`, nor does it support `1.4 s`, nor does it support languages that use non-ascii characters.

File: src/main/resources/lang/default.lang [hunk 3]
```diff
@@ -343,17 +343,31 @@ tree types:
...
-	minute: minute¦s
-	hour: hour¦s
-	day: day¦s
-	week: week¦s
-	month: month¦s
-	year: year¦s
-	real: real, rl, irl
-	minecraft: mc, minecraft
+  millisecond: millisecond¦s
+  tick: tick¦s
```
Comment (sovdeeth) [inline src/main/resources/lang/default.lang:347]: milliseconds and ticks can no longer be parsed since they have no full or short name.

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/util/Timespan.java (hunk 1 above)
Comment (Moderocky) [inline src/main/java/ch/njol/skript/util/Timespan.java:216]: This should be a field since it doesn't need to be done multiple times.

Since we have all of the short forms of the time periods, you could actually just make a regex searching for those rather than any combination of random letters.

**Review by Moderocky — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/util/Timespan.java [hunk 4]
```diff
@@ -159,31 +177,68 @@ else if (length == 3 && !hasMs || length == 4) // HH:MM:SS[.ms]
...
+
+				totalMillis += Math.round(amount * millis);
+
 				isMinecraftTimeSet = true;
-				
 			}
 		}
 
-		return new Timespan(t);
+		return totalMillis > 0 ? new Timespan(totalMillis) : null;
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/util/Timespan.java:197]: this is wrong, what happens to `0 seconds`

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/test/skript/tests/misc/timespans.sk [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	assert {_timespan} is a timespan with "Failed to parse '1 mo, 30 days, 12w, 20 minutes and 15s'"
+	assert {_timespan} is 1 month, 30 days, 12 weeks, 20 minutes and 15 seconds with "Failed to set variable to '1 mo, 30 days, 12w, 20 minutes and 15s'"
+
+	set {_timespan} to 50 minutes and 45 seconds
+	assert {_timespan} is a timespan with "Failed to parse '50 minutes and 45 seconds'"
+	assert {_timespan} is 50m and 45s with "Failed to set variable to '50 minutes and 45 seconds'"
+
+	# edge cases
+	set {_timespan} to 0.5d
+	assert {_timespan} is a timespan with "Failed to parse '0.5d'"
```
Comment (sovdeeth) [inline src/test/skript/tests/misc/timespans.sk:42]: technically these cases are not very useful since the following assert would catch it anyway

File: src/test/skript/tests/misc/timespans.sk [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+	assert {_timespan} is 0.5d with "Failed to set variable to 0.5d"
```
Comment (sovdeeth) [inline src/test/skript/tests/misc/timespans.sk:43]: try not to check with the exact same thing you set it to. If there's a value issue with 0.5d, this wouldn't catch it because 0.5d = 0.5d no matter what 0.5d is. This goes for all the tests.

It'd also be good to just do some simple `add x to y` /remove/clear tests while you're here. Maybe check for overflows too.

and finally some decimal tests with the long names and with multiple units (1.5 days and 2.5 seconds)

**Review by sovdeeth — CHANGES_REQUESTED** [review]

**Review by sovdeeth — APPROVED** [review]

**Review by Moderocky — COMMENTED** [review]
The code looks ok, but I'm a bit concerned about allowing this outside commands because it feels like it could be weird and ambiguous in code.
My concern is mainly that `m` is the international standard abbreviation for metres (a unit of measurement for any americans out there), I can guarantee this will trip up some beginners doing things like `the block 20m in front of player` and wondering why it's not working.
I think it's also worth noting that Minecraft uses `d` and `s` as number type tokens in its stringified NBT format (and in fact some real languages use `D` as a double token) so I can see somebody getting confused by `0.5d` or whatever, but I think that's a much lesser concern.

If the others are happy with this then I'm fine with it, but if it was me I would probably just restrict this to commands or text -> timespan parsing where a shorthand (e.g. 1h30m, 3d2y) feels more appropriate.

**Comment by Asleeepp** [comment]
> The code looks ok, but I'm a bit concerned about allowing this outside commands because it feels like it could be weird and ambiguous in code. My concern is mainly that `m` is the international standard abbreviation for metres (a unit of measurement for any americans out there), I can guarantee this will trip up some beginners doing things like `the block 20m in front of player` and wondering why it's not working. I think it's also worth noting that Minecraft uses `d` and `s` as number type tokens in its stringified NBT format (and in fact some real languages use `D` as a double token) so I can see somebody getting confused by `0.5d` or whatever, but I think that's a much lesser concern.
> 
> If the others are happy with this then I'm fine with it, but if it was me I would probably just restrict this to commands or text -> timespan parsing where a shorthand (e.g. 1h30m, 3d2y) feels more appropriate.

The original PR was locked behind command context, but i removed that requirement with this new PR, as I personally felt that locking this behind just commands felt a little bit silly, but you do have some valid arguments, if anyone else agrees with Kenzie on this, i'll lock it behind commands only

**Comment by Moderocky** [comment]
> if anyone else agrees with Kenzie on this, i'll lock it behind commands only

I'm not like totally against it, I just feel like it's the sort of thing that might bite us down the line, would be interested to see others' opinions :smile:

**Comment by sovdeeth** [comment]
> > if anyone else agrees with Kenzie on this, i'll lock it behind commands only
> 
> I'm not like totally against it, I just feel like it's the sort of thing that might bite us down the line, would be interested to see others' opinions 😄

Yes i think it'd be better as command context only. I understand the need for brevity in commands (and it's less likely to cause issues there), but in actual code i think writing out the units makes much more sense, makes code more readable, prevents possible future conflicts, etc.

File: src/test/skript/tests/misc/timespans.sk [hunk 7]
```diff
@@ -1,84 +1,50 @@
...
-	assert {_timespan} is a timespan with "Failed to parse large timespan"
-	assert {_timespan} is greater than 9999999998 days with "Failed to handle large timespan correctly"
+	execute command "/testtimespan invalidinput"
+	assert {timespan} is "invalid" parsed as timespan to fail with "%{timespan}% was incorrectly parsed!"
 
+	execute command "/testtimespan 1mo 2w 3d 4h 5m 6s"
+	assert {timespan} is "1 month, 2 weeks, 3 days, 4 hours, 5 minutes and 6 seconds" parsed as timespan with "%{timespan}% doesn't equal 1 month, 2 weeks, 3 days, 4 hours, 5 minutes and 6 seconds!"
 
+	execute command "/testtimespan 9999999999d"
+	assert {timespan} is "9999999999 days" parsed as timespan with "%{timespan}% doesn't equal 9999999999 days!"
```
Comment (Asleeepp) [inline src/test/skript/tests/misc/timespans.sk:50]:
```suggestion
	assert {timespan} is "9999999999 days" parsed as timespan with "%{timespan}% doesn't equal 9999999999 days!"
	
```

**Review by Asleeepp — COMMENTED** [review]
dont really feel like going back into intellij to fix this

**Review by Moderocky — APPROVED** [review]
Looks good :)
